src/gululog_idx.erl

```diff
@@ -285,39 +315,172 @@ truncate(Dir, #idx{tid = Tid, fd = Fd} = Idx, SegId, LogId, BackupDir) ->
+  Fd = open_reader_fd(Dir, SegId),
   try
     {ok, <<Version:8>>} = file:read(Fd, 1),
-    scan_locate_per_vsn(Fd, SegId, LogId, Version)
+    ReadFun = fun(LogId) -> read_file_entry(Version, Fd, SegId, LogId) end,
+    find_first_logid_since_in_file(ReadFun, LogIdLo, LogIdHi, Ts)
   after
     file:close(Fd)
   end.
 
--spec scan_locate_per_vsn(file:fd(), segid(), logid(), logvsn()) ->
-        {segid(), position()}.
-scan_locate_per_vsn(Fd, SegId, LogId, 1) ->
-  %% The offset caculate by per-entry size + one byte version
-  Location = (LogId - SegId) * ?FILE_ENTRY_BYTES_V1 + 1,
-  {ok, <<FileEntry:?FILE_ENTRY_BITS_V1>>} =
-    file:pread(Fd, Location, ?FILE_ENTRY_BYTES_V1),
-  ?ETS_ENTRY(SegId, LogId, Position) = ?FROM_FILE_ENTRY_V1(SegId, FileEntry),
-  {SegId, Position}.
+%% @private Read .idx file entries to locate the first logid since the given timestamp.
+-spec find_first_logid_since_in_file(fun((logid()) -> entry()),
+                                     logid(), logid(), os_sec()) -> logid().
+find_first_logid_since_in_file(_ReadFun, LogId, LogId, _Ts) ->
+  LogId;
+find_first_logid_since_in_file(ReadFun, LogIdLo, LogIdHi, Ts) ->
+  case ReadFun((LogIdLo + LogIdHi) bsr 1) of
+    ?ENTRY(LogIdMi, TsX, _SegId, _Pos) when TsX >= Ts ->
+      find_first_logid_since_in_file(ReadFun, LogIdLo, LogIdMi, Ts);
+    ?ENTRY(LogIdMi, TsX, _SegId, _Pos) when TsX < Ts ->
+      find_first_logid_since_in_file(ReadFun, LogIdMi+1, LogIdHi, Ts)
+  end.
+
+%% @private Read the latest (last) log entry of the given segment.
+-spec read_latest_in_seg(dirname(), index(), segid()) -> false | entry().
+read_latest_in_seg(_Dir, #idx{tid = Tid, segid = SegId}, SegId) ->
+  %% latest entry is never deleted from cache
+  last_in_cache(Tid);
+read_latest_in_seg(Dir, #idx{tid = Tid, segid = LatestSegId}, SegId) ->
+  not is_out_of_range(Tid, SegId) andalso
+  read_entry(Dir, Tid, find_latest_logid_in_seg(Tid, SegId, LatestSegId)).
+
+%% @private Search backward from a greater segment id to find the last logid of the given segment.
+-spec find_latest_logid_in_seg(cache(), segid(), segid()) -> logid().
+find_latest_logid_in_seg(Tid, SegId, SegIdX) ->
+  true = (SegIdX > SegId), %% assert
+  LogId = (SegId + SegIdX) bsr 1,
+  find_latest_logid_in_seg(Tid, SegId, SegIdX, LogId).
+
+-spec find_latest_logid_in_seg(cache(), segid(), segid(), logid()) -> logid().
+find_latest_logid_in_seg(_Tid, _SegId, SegIdX, SegIdX) ->
+  SegIdX - 1;
+find_latest_logid_in_seg(Tid, SegId, SegIdX, LogId) ->
+  case get_segid(Tid, LogId) of
+    SegId ->
+      %% test a greater logid
+      NewLogId = (LogId + 1 + SegIdX) bsr 1,
+      find_latest_logid_in_seg(Tid, SegId, SegIdX, NewLogId);
+    SegIdY ->
+      find_latest_logid_in_seg(Tid, SegId, SegIdY)
+  end.
+
+%% @private Get segid for the given logid.
+get_segid(Idx, LogId) ->
+  case lookup_cache(Idx, LogId) of
+    false                  -> get_segid(Idx, prev_logid(Idx, LogId));
+    ?ENTRY(_, _, SegId, _) -> SegId
+  end.
```

Comment:
1, no `-spec`
2, there is ambiguity for argument `Idx`

```
-spec get_segid(cache(), logid()) -> segid().
get_segid(Tid, LogId) ->
  case lookup_cache(Tid, LogId) of
    false                  -> get_segid(Tid, prev_logid(Tid, LogId));
    ?ENTRY(_, _, SegId, _) -> SegId
  end.
```
maybe will better ?